plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/dataflow/guards/GuardType.java

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (c) 2018 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.flowgraphs.dataflow.guards;
+
+import org.eclipse.n4js.n4JS.Expression;
+
+/**
+ * Type of a guard.
+ */
+public enum GuardType {
+	/**
+	 * For {@link Expression}s that check for null:<br/>
+	 * {@code if (a == null)}
+	 */
+	IsNull,
+	/**
+	 * For {@link Expression}s that check for undefined:<br/>
+	 * {@code if (a == undefined || a == void 0 || typeof a == "undefined")}
+	 */
+	IsUndefined,
+	/**
+	 * For {@link Expression}s that check for truthy:<br/>
+	 * {@code if (a)}
+	 */
+	IsTruthy,
+	/**
+	 * For {@link Expression}s that check for 0:<br/>
+	 * {@code if (a == 0)}
+	 */
+	IsZero,
+	/**
+	 * For {@link Expression}s that check for a type:<br/>
+	 * {@code if (a instanceof AType)}
+	 */
+	InstanceOf,
```

Comment:
as discussed: we should consider moving things like ``isNull``, ``isUndefined``, and ``instanceOf`` and the related logic from the engine to client code (possible draw back: performance?). But this refactoring can be considered when we have more experience with concrete analyses (e.g. type guards).

Reply:
True, this could be refactored to be extensible for client analyses later.